sdk/mobile-center-distribute/src/main/java/com/microsoft/azure/mobile/distribute/Distribute.java

```diff
@@ -1119,6 +1127,39 @@ private void viewReleaseNotes(ReleaseDetails releaseDetails) {
     }
 
     /**
+     * Store failed package info hash in preferences.
+     */
+    private synchronized void storeUpdateSetupFailedPackageHash() {
+        if (mPackageInfo != null && mUpdateSetupFailedDialog != null) {
+            PreferencesStorage.putString(PREFERENCE_KEY_UPDATE_SETUP_FAILED_PACKAGE_HASH_KEY, DistributeUtils.computeReleaseHash(mPackageInfo));
+        } else {
+            showDisabledToast();
+        }
+    }
+
+    /**
+     * Redirect user to install url page in browser and clear current failed package hash from preferences.
+     */
+    private synchronized void handleUpdateFailedDialogReinstallAction() {
+        if (mUpdateSetupFailedDialog != null) {
```

Comment:
prefer passing dialog as parameter and test against that instead of null like in other places